docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+    GangScheduler *kubeflowv1.GangScheduler `json:"gangScheduler,omitempty"`
+}
+
+// One of the specs can be selected.
+type MLSpec struct {
+
+    // Custom Spec for Torch
+    TorchSpec *TorchSpec `json:"torchSpec,omitempty"`
+
+    // Custom Spec for MPI
+    MPISpec *MPISpec `json:"mpiSpec,omitempty"`
+}
+```
+
+### The Gang Scheduler API
+
+Gang scheduler plugin is used to create appropriate `PodGroup` for Volcano or scheduler plugins.
+
+```golang
+type GangScheduler struct {
+    // Plugin for gang scheduling.
+    Plugin *GangSchedulerPlugin `json:plugin,omitempty"`
+
+    // Time threshold to schedule PodGroup for gang scheduling.
+    ScheduleTimeoutSeconds string `json:scheduleTimeoutSeconds,omitempty"`
+}
+
+type GangSchedulerPlugin string
+
+const (
+    GangSchedulerPluginVolcano GangSchedulerPlugin = "volcano"
+    GangSchedulerPluginSP      GangSchedulerPlugin = "scheduler-plugins"
+)
+```
+
+### The Torch Spec API
+
+The `TorchSpec` API represents the configuration for the PyTorch distributed training. This configuration
+allows platform engineers to explicitly configure `torchrun` setting.
+
+The distributed parameters are taken from the
+[PyTorch distributed launch run](https://github.com/pytorch/pytorch/blob/94dc3253a0fefbfb95fbe467ddd628e4c2eb08d7/torch/distributed/run.py).
+
+For Elastic Training we will always pass the following parameters:
+
+```bash
+--rdzv-backend=c10d
+
+--rdzv-id will be set automatically.
+
+--rdzv-endpoint will always point to the node-0 Pod.
+```
+
+Since the [etcd and etcd-v2 are legacy rendezvous](https://pytorch.org/docs/stable/elastic/run.html#note-on-rendezvous-backend),
+we won't support them in `TorchSpec`. We can introduce them in the future if users will require them.
+
+```golang
+// TorchSpec represents the configuration for PyTorch.
+type TorchSpec struct {
+
+    // Number of Procs per Node.
+    NumProcPerNode int `json:"numProcPerNode,omitempty"`
+
+    // Used for single-node multi-worker training
+    Standalone bool `json:"standalone,omitempty"`
+
+    // Torch Elastic Policy.
+    ElasticPolicy *TorchElasticPolicy `json:"elasticPolicy,omitempty"`
+}
+
+// If the Elastic Policy is set, the numNodes parameter is ignored.
+// --nnodes=minNodes:maxNodes
+type TorchElasticPolicy struct {
+
+    // The limits to restart TrainJob.
+    // Insert it to the JobSet.spec.failurePolicy.maxRestarts
+    MaxRestarts *in32 `json:"maxRestarts,omitempty"`
+
+    // Min number of nodes for HPA and torchrun.
+    MinNodes *in32 `json:"minNodes,omitempty"`
+
+    // Max number of nodes for HPA and torchrun.
+    MaxNodes *in32 `json:"maxNodes,omitempty"`
+
+    // Metrics for scale up and down replicas.
+    Metrics []autoscalingv2.MetricSpec `json:"metrics,omitempty"`
+}
+```
+
+### The MPI Spec API
+
+The `MPISpec` API represents the configuration for training using MPI orchestration.
+E.g. creation of host-files and SSH keys. Using MPI might be more efficient for training on HPC
```

Comment:
HPC clusters will usually have a workload manager that handles the MPI bootstrap and gets rid of the need for a hostfile and ssh. For example, flux uses zeromq for communication and already is aware of its hosts. One thing I want to keep in mind is, because we are anticipating running user space Kubernetes on our systems for the future, we want these abstractions to map nicely. For that I would say to leave this open and flexible to support not just different MPI interface commands, but also entirely different workload managers. As a side note, I am very happy to see MPI here :)

Reply:
Sure, I think, we can support other workload managers in our TrainingRuntime in the future (e.g. Flux, Slurm).
For example, I know that right now running PMIx is not possible with MPI Operator: https://github.com/kubeflow/mpi-operator/issues/12.

@vsoch Any thoughts on how we can design our `MLSpec` to make it more generic for other workload managers ?

Reply:
To step back - what is distinct about an MPISpec? Right now it is the launcher, and need to setup ssh? If we were to deploy a workload manager, and one that didn't require a special launcher node, the resources could be equivalent, and the main difference would be that one node (the leader) is running a slightly different command than the workers (followers). I think what we should do is write down explicitly the features that are unique to an MPISpec, figure out the generic design, and try to remove needing to name specific implementations to the largest degree possible. Arguably, an MPISpec is valid for "something that uses MPI" but should be flexible to change in that landscape over time (and even supporting the workload managers).

Reply:
The main motivation behind `MLSpec` is to simplify experience for Platform Engineers on how to configure various parameters for different runtimes. For example, `ElasticPolicy` parameter is only applicable for PyTorch since TensorFlow distributed doesn't support it.
For the workload managers, I am wondering do they (Slurm, Flux, and others) are always use MPI to run distributed job, or we can run them without MPI. For example, Slurm + `torchrun`. cc @kuizhiqing

Reply:
Yes, of course you can run non MPI jobs under a workload manager! I didn't use MPI for my entire graduate school career (it was more machine learning and biosciences pipelines). You can use the same features (the ability to launch and schedule with batch, or to sub-instances in a graph) in absence of MPI. As as example, I've run torchrun with the Flux Operator ([example](https://github.com/flux-framework/flux-operator/tree/6a223538519adafd2785ebe4cd92afc6d34d6444/examples/machine-learning/pytorch)). 

Workload managers are not explicitly wed to MPI, but because MPI is a bread and butter HPC communication interface ([detail](https://www.osti.gov/servlets/purl/1860774), it's standard that the workload managers support running these apps optimally (low latency, bootstrap, etc). As a side note on elasticity, we have support for a faux kind of elasticity for the Flux Operator (adding nodes dynamically) and should have full support in bare metal flux (without Kubernetes) to grow and shrink jobs soon.

Reply:
That makes sense, so in that case maybe in the future we can introduce another union type in our MLSpec to support it (e.g. FluxSpec, SlurmSpec).
That gives Platform Engineers ability to create training blueprints utilising various workload managers capabilities.

Reply:
Absolutely. I can offer to help with that, when the time comes.